### DISCUSSION TITLE
How is snapshot assembled in fom.solve

### DISCUSSION THREAD

**Discussion by jacktang** [pymor/pymor#2009] [Q&A]
Hello There!

I want to compare pod combining neural network method and dmd (mainly pydmd). And here are two questions:

1. How is snapshot assembled in `fom.solve()`, suppose the problem is time dependent and the space is 3d geometry. Is it compatible to dmd snapshot?
2. How about the dmd module in pymor? Is it mature enough?

Thanks a lot!

**Comment by pmli** [comment]
Hi @jacktang.

> 1. How is snapshot assembled in `fom.solve()`, suppose the problem is time dependent and the space is 3d geometry. Is it compatible to dmd snapshot?

It is expected that `solve` returns a `VectorArray` containing the solution trajectory (computed using a time stepper). The [`dmd`](https://docs.pymor.org/2022-2-1/autoapi/pymor/algorithms/dmd/index.html) algorithm in pyMOR accepts any `VectorArray`. I suspect that PyDMD expects snapshots as a NumPy array, for which you can use the `to_numpy` method of the `VectorArray` (if available).

> 2. How about the dmd module in pymor? Is it mature enough?

There are some basic variants of DMD implemented in [`pymor.algorithms.dmd`](https://github.com/pymor/pymor/blob/54ff5159ad5aaa8c51a084490d8de8372418d355/src/pymor/algorithms/dmd.py). I believe that adding more variants would be in the scope of pyMOR.

**Reply by jacktang** [reply]
Hello @pmli 
> It is expected that `solve` returns a `VectorArray` containing the solution trajectory (computed using a time stepper). The [`dmd`](https://docs.pymor.org/2022-2-1/autoapi/pymor/algorithms/dmd/index.html) algorithm in pyMOR accepts any `VectorArray`. I suspect that PyDMD expects snapshots as a NumPy array, for which you can use the `to_numpy` method of the `VectorArray` (if available).
> 
Thanks for you reply! AFAIK, in dmd methodology, two-dimensional or three-dimensional vector field data at time $t$ is flattened into a single tall column vector, and the final data matrix is a tall and skinny matrix. Does `fom.solve()` organize the snapshot in the same way? I try to inspect the thermal block problem's solution matrix, but I don't know how to combine the value in solution matrix with spatial information. 

Regards
/Jack

**Reply by pmli** [reply]
> AFAIK, in dmd methodology, two-dimensional or three-dimensional vector field data at time t is flattened into a single tall column vector, and the final data matrix is a tall and skinny matrix. Does `fom.solve()` organize the snapshot in the same way?

A `VectorArray` just needs to behave like a sequence of vectors. This can be 1D NumPy arrays arranged in a 2D NumPy array as done in `NumpyVectorArray`, but it can be anything else that implements methods of the `VectorArray`. The implementation of DMD in pyMOR also doesn't assume the form of the vectors, it only uses the `VectorArray` interface.

> I try to inspect the thermal block problem's solution matrix, but I don't know how to combine the value in solution matrix with spatial information.

`VectorArrays` have the [`dofs`](https://docs.pymor.org/2022-2-1/autoapi/pymor/vectorarrays/interface/index.html#pymor.vectorarrays.interface.VectorArray.dofs) method (if implemented), which returns the value of the vectors for given indices. I'm not sure if it is possible generally in pyMOR to connect the spatial coordinates with vector indices.
